lib/logstash/filters/kv.rb

```diff
@@ -238,27 +239,39 @@ def filter(event)
     # Add default key-values for missing keys
     kv = @default_keys.merge(kv)
 
-    # If we have any keys, create/append the hash
-    if kv.length > 0
-      if @target.nil?
-        # Default is to write to the root of the event.
-        dest = event.to_hash
-      else
-        if !event[@target].is_a?(Hash)
-          @logger.debug("Overwriting existing target field", :target => @target)
-          dest = event[@target] = {}
-        else
-          dest = event[@target]
-        end
+    unless kv.empty?
+      if @target
+        event.remove("#{@target}")
+        @logger.debug? && @logger.debug("Overwriting existing target field", :target => @target)
       end
 
-      dest.merge!(kv)
+      inject_fields(event, @target, kv)
+
       filter_matched(event)
     end
   end
 
   private
 
+  # recursiverly inject the data Hash content into event. the data parameter must be Hash on initial call.
+  #
+  # this is important to do here because we cannot assume in the future the in-place mutability of the Event#to_hash structure.
+  # TODO: (colin) we can certainly optimize this by introducing a new Event API to directly set fields without going through the
```

Comment:
ah, yes, that's because the `append` method happens to also works with a `Hash` argument but it was meant to work with an `Event` argument. so it's not a good idea to use append, and the Java Event implementation correctly validates the argument ...